### PR TITLE
Grafana-CLI: Wrapper for `grafana-cli` within RPM/DEB packages and config/homepath are now global flags

### PR DESCRIPTION
**Pull request by gotjosh**
**What this PR does / why we need it**:

When our users install the *nix packed version of grafana (via RPM or deb), tendency is to use the services and scripts installed as part of the package for grafana-server. These leverage the default configuration options by specifying several default paths.

This introduces a similar approach for the grafana-cli binary. We expose it through a wrapper to ensure a proper configuration is in place. 

To enable that, we tuck away into the `/usr/share/grafana` (default grafana home) directory the original binary (grafana-cli) and then use a bash script named grafana-cli as the wrapper.

In addition to that, this PR also:

- Makes the `--config` and `--homepath` flags global. Which allow us to significantly simplify the wrapper logic. 
- Adds a new  `--optionsString` which allow us to override default configuration options as a string. A thing to note is that this was and is still supported by passing these arguments as trailing args to the command, with the particularity that the trailing arguments take precedence over the flag to allow for backward compatibility. That means:

```bash
$ ./bin/darwin-amd64/grafana-cli -d --optionsString="cfg:default.paths.data=/dev/nullX cfg:default.paths.logs=/dev/nullX" --config=/dev/null admin reset-admin-password 1234 cfg:default.paths.plugins=/dev/nullY cfg:default.paths.logs=/dev/nullY
INFO[06-20|14:33:00] Config loaded from                       logger=settings file=/Users/josueabreu/go/src/github.com/grafana/grafana/conf/defaults.ini
INFO[06-20|14:33:00] Config loaded from                       logger=settings file=/dev/null
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.data=/dev/nullX"
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.logs=/dev/nullY"
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.plugins=/dev/nullY"
INFO[06-20|14:33:00] Path Home                                logger=settings path=/Users/josueabreu/go/src/github.com/grafana/grafana
INFO[06-20|14:33:00] Path Data                                logger=settings path=/dev/nullX
INFO[06-20|14:33:00] Path Logs                                logger=settings path=/dev/nullY
INFO[06-20|14:33:00] Path Plugins                             logger=settings path=/dev/nullY
INFO[06-20|14:33:00] Path Provisioning                        logger=settings path=/Users/josueabreu/go/src/github.com/grafana/grafana/conf/provisioning
INFO[06-20|14:33:00] App mode production                      logger=settings
INFO[06-20|14:33:00] Connecting to DB                         logger=sqlstore dbtype=sqlite3
INFO[06-20|14:33:00] Starting DB migration                    logger=migrator

Error: New password is too short

NAME:
   Grafana cli admin reset-admin-password - reset-admin-password <new password>

USAGE:
   Grafana cli admin reset-admin-password [arguments...
```
- Logs the paths configured as part of the command run to provide some feedback for the user if the `--debug` global flag is used

### CODE REVIEW

**Comment by gotjosh** [comment]
Superseds #17596

**Comment by torkelo** [comment]
This looks like a nice improvement. 

We would need a bat/windows version as well.

**Comment by gotjosh** [comment]
A bit of an update from me: Based on a conversation I had with @xlson, we were both under the impression that we currently have support for "muting" specific loggers based on this [configuration flag](https://github.com/grafana/grafana/blob/master/conf/sample.ini#L393). However, this is not the case - what this configuration does is supersed whatever level you originally specify on a particular logger.

**Comment by gotjosh** [comment]
A follow up from my message above; I've added the option to mute logs as part of #17710

This PR now depends on that to silence the logs - the part that silences logs in debug only for the CLI is ready but will merge #17710 first before I push that code to this PR.

**Comment by torkelo** [comment]
I can't see how this depends on the mute PR, how and why?

**Comment by marefr** [comment]
> we were both under the impression that we currently have support for "muting" specific loggers based on this configuration flag. However, this is not the case - what this configuration does is supersed whatever level you originally specify on a particular logger.

Guess this is an existing problem with the cli (logs are being written of services starting up etc). I would recommend open a separate issue for resolving that if it's not critical for the success of this feature/fix.

**Comment by torkelo** [comment]
Personally I think the cli should not connect to DB or use internal grafana-server config & services. It should talk to grafana-server via the HTTP api. 

Having the cli tool include big parts of the grafana-server runtime & config, and DB code feels very wrong to me. Wonder how other systems cli tools do this?

**Comment by gotjosh** [comment]
Removed my comment re the dependency. Apologies for falling into a wrong assumption and deviating everyone. I'm with @marefr on this one, I'll create an issue for any other follow-ups.

**Comment by gotjosh** [comment]
@torkelo I don't have any strong opinions either way. They both have trade-offs. 

I've seen both approaches work with great success in the wild although I have to admit that the latter (talk over HTTP) is more common these days in Go (due to the ability to test/integrate a bit easier).

File: build.go [hunk 1]
```diff
@@ -323,10 +328,16 @@ func createPackage(options linuxPackageOptions) {
...
 	runPrint("mkdir", "-p", filepath.Join(packageRoot, "/usr/sbin"))
 
-	// copy binary
-	for _, binary := range binaries {
-		runPrint("cp", "-p", filepath.Join(workingDir, "tmp/bin/"+binary), filepath.Join(packageRoot, "/usr/sbin/"+binary))
-	}
+	// The grafana-cli binary is exposed through a wrapper to ensure a proper
+	// configuration is in place. To enable that, we need to store the original
+	// binary in a separate location to avoid conflicts.
+	runPrint("cp", "-p", filepath.Join(workingDir, "tmp/bin/"+cliBinary), filepath.Join(packageRoot, options.homeDir, cliBinary))
```
Comment (xlson) [inline build.go]: How about adding a `bin` folder to the home folder and putting it in there? Seems a bit cleaner to me.

File: pkg/cmd/grafana-cli/main.go [hunk 2]
```diff
@@ -51,6 +51,18 @@ func main() {
 			Name:  "debug, d",
 			Usage: "enable debug logging",
 		},
+		cli.StringFlag{
```
Comment (xlson) [inline pkg/cmd/grafana-cli/main.go]: How about calling this `configOverrides` or something like that instead? Just `options` could work as well.

**Review by xlson — CHANGES_REQUESTED** [review]
Looks great. Two minor things that me may want to change to. What do you think?

File: pkg/cmd/grafana-cli/main.go (hunk 2 above)
Comment (gotjosh) [reply pkg/cmd/grafana-cli/main.go]: Sounds good to me.

File: build.go (hunk 1 above)
Comment (gotjosh) [reply build.go]: Sounds good to me.

**Review by xlson — APPROVED** [review]
LGTM!

**Comment by gotjosh** [comment]
A head's up @xlson that I have removed the commit that moves the binary into a `/bin` folder. When I manually tested the change it doesn't seem to be creating the folder nor moving the binary which ends up erroring per the wrapper safeguard. 

We want to push this feature out asap as part of the new release and I believe the change to be non-critical. I have documented that change and two others we talked about as part of https://github.com/raintank/hosted-grafana/pull/96 to execute as a follow-up.

**Comment by gotjosh** [comment]
Managed to find and fix the problem @xlson. Merging this in.

**Pull request merged**